<img src="http://lssds.aura-astronomy.org/winter_school/sites/default/files/sods_atfdds_header01.jpg" alt="La Serena School for Data Science: Applied Tools for Data-driven Sciences" />


# 1. Introduction to Docker Compose:

Docker Compose is a tool that enables you to define and manage multi-container Docker applications. It allows you to describe the entire application stack, including multiple services, networks, and volumes, in a single YAML file. This theoretical introduction will help you understand the fundamental concepts of Docker Compose.

<img src="https://tcude.net/content/images/2022/01/MainImage-2.jpeg" alt="La Serena School for Data Science: Applied Tools for Data-driven Sciences" />

## 1.1. Simplifying Multi-Container Applications:
---

- Many modern applications are composed of multiple services, each running in its own container. These services often need to communicate and interact with each other.
- Docker Compose simplifies the process of managing and coordinating these multi-container applications by allowing you to define and control the entire stack with a single configuration file.

## 1.2. Docker Compose YAML File:
___

- A Docker Compose YAML file (***docker-compose.yml***) is used to define the components of a multi-container application. It specifies services, networks, volumes, and other configurations.
- The YAML file is human-readable and allows you to define the relationships, dependencies, and settings for each service in your application stack.

## 1.3. Services:
___
- A service is a unit of an application defined in the Docker Compose YAML file. It corresponds to a container and defines how the container should be configured and run.
- Each service in the ***docker-compose.yml*** file can have its own image, environment variables, ports, volumes, and other settings.

## 1.4. Networks:
___
- Docker Compose allows you to create custom networks for your services. Networks enable communication between services while isolating them from external traffic.
- Services within the same network can communicate using service names as hostnames.

## 1.5. Volumes:
___
- Volumes in Docker Compose allow you to persist and share data between containers. They are used to store data that needs to persist beyond the container's lifecycle.
- Volumes are particularly useful for database containers or containers with important configuration files.

## 1.6. Advantages of Docker Compose:
___
- **Simplicity:** Docker Compose abstracts away the complexity of managing multiple containers, making it easier to define and maintain complex application stacks.
- **Reproducibility:** With a single docker-compose.yml file, you can define the entire application stack, ensuring consistency across environments.
- **Portability:** Docker Compose configurations can be versioned and shared, allowing teams to collaborate and ensure that the same application runs consistently everywhere.
- **Isolation:** Services in Docker Compose run in isolated containers, providing process-level isolation while allowing seamless communication.

## 1.7. Use Cases:
___
- Docker Compose is particularly useful for development and testing environments, where you need to quickly spin up an entire application stack for local testing or debugging.
- It's also valuable for creating repeatable setups for continuous integration and deployment pipelines.

## 1.8 Verify Docker Compose Installation:
___
After installing Docker Desktop, you should have Docker Compose automatically available. To verify the installation, open a terminal and run the following command:



In [17]:
!docker-compose --version

Docker Compose version v2.17.2


## Summary: 
___
Docker Compose simplifies the management of multi-container applications by providing a declarative way to define and orchestrate services, networks, and volumes. It's a valuable tool for both developers and DevOps teams to streamline the deployment and testing of complex applications.

## 2. Run Your First Docker Container:
___


### Step 1: Create a Directory for Your Project:
Open a terminal and create a new directory for your Dask cluster project:

In [ ]:
ls

In [ ]:
!mkdir dask-cluster

In [ ]:
!cd dask-cluster

### Step 2: Create a docker-compose.yml File:

Inside your project directory, create a file named **docker-compose.yml** using a text editor of your choice. This file will define your Dask cluster configuration.

### Step 3: Define Services in docker-compose.yml:

Add the following content to your docker-compose.yml file:

In [18]:
!cat dask-cluster/docker-compose.yml

version: '3'
services:
  scheduler:
    image: daskdev/dask:latest
    command: ["dask-scheduler"]
    ports:
      - "8786:8786"

  worker:
    image: daskdev/dask:latest
    command: ["dask-worker", "scheduler:8786"]
    depends_on:
      - scheduler

This configuration defines two services: **scheduler** and **worker**.

- The ***scheduler*** service uses the ***daskdev/dask*** image and starts the Dask scheduler using the dask-scheduler command. It exposes port 8786 for communication.
- The ***worker*** service also uses the same image and starts a Dask worker using the ***dask-worker*** command. It depends on the ***scheduler*** service and connects to the scheduler's port 8786.

### Step 4: Start the Dask Cluster:

In your terminal, navigate to your project directory and run the following command to start the Dask cluster:

In [19]:
!docker ps

CONTAINER ID   IMAGE     COMMAND   CREATED   STATUS    PORTS     NAMES


In [ ]:
!docker-compose up

Run the docker-compose up command with the -f or --file option followed by the path to your docker-compose.yml file.

In [20]:
!docker-compose -f dask-cluster/docker-compose.yml up

[+] Running 2/0
 ✔ Container dask-cluster-scheduler-1  Created                             0.0s 
 ✔ Container dask-cluster-worker-1     Created                             0.0s 
Attaching to dask-cluster-scheduler-1, dask-cluster-worker-1
dask-cluster-scheduler-1  | no environment.yml
dask-cluster-scheduler-1  | + '[' '' ']'
dask-cluster-scheduler-1  | + '[' '' == true ']'
dask-cluster-scheduler-1  | + CONDA_BIN=/opt/conda/bin/conda
dask-cluster-scheduler-1  | + '[' -e /opt/app/environment.yml ']'
dask-cluster-scheduler-1  | + echo 'no environment.yml'
dask-cluster-scheduler-1  | + '[' '' ']'
dask-cluster-scheduler-1  | + '[' '' ']'
dask-cluster-scheduler-1  | + exec dask-scheduler
dask-cluster-worker-1     | no environment.yml
dask-cluster-worker-1     | + '[' '' ']'
dask-cluster-worker-1     | + '[' '' == true ']'
dask-cluster-worker-1     | + CONDA_BIN=/opt/conda/bin/conda
dask-cluster-worker-1     | + '[' -e /opt/app/environment.yml ']'
dask-cluster-worker-1     | + echo 'no enviro

Docker Compose will pull the necessary images and start the scheduler and worker containers

### Step 5: Check the cluster:



In [ ]:
!docker ps

### Step 6: Stop the Cluster:

To stop the Dask cluster, press **Ctrl+C** in the terminal where you started the cluster with docker-compose up.

### Step 7: Clean Up:

If you want to remove the containers and network created by Docker Compose, run the following command in your project directory:

In [21]:
!docker-compose down

no configuration file provided: not found


### Summary:
That's it! You've created a Docker Compose configuration for a Python Dask cluster. This setup allows you to easily manage and scale your Dask cluster for distributed computing tasks.